platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,942 @@
+
+**A note on cloning children**
+To pass the form state to `Page`'s children, it will clone all its children and
+pass the form state props along with the child's given props. Implementation
+will look something like the following:
+```js
+const newChildren = React.Children.map(
+  this.props.children,
+  child => React.cloneElement(child, { formData, pathParams }));
+```
+
+This means developers can create form pages in a couple different ways. 
+
+_Using the form state:_
+```jsx
+<Page>
+  {({ formData, pathParams }) => (/* Use the form state to render the correct fields */)}
+</Page
+```
+
+_Not using the form state:_
+```jsx
+<Page>
+  <StringField name="foo" />
+  <StringField name="bar" />
+</Page
+```
+
+In the above example, both `StringField`s will have the following props:
+- `name`
+- `formData`
+- `pathParams`
+
+The form state props `formData` and `pathParams` will simply go unused.
+
+##### `Route` Component
+**Purpose:**
+The `Route` will be re-exported from `react-router-dom`. The purpose of
+re-exporting is twofold:
+- Simplifies imports
+- Ensures only one version of `react-router-dom` is used
+
+**Usage requirements:**
+- `Route` may only be a child of `Router`
+
+**Why not just use `Page`?**
+The `Chapter` and `Page` components are meant to be used to define pages where
+the user may enter data. These pages are sequential and have back and continue
+buttons. `Route`s are used for pages in the application that live outside that
+form flow. These pages may include:
+- `/introduction`
+- `/form-saved`
+- `/error`
+
+##### Functionality: Array pages
+Array pages are one or more pages that are looped through in a form flow, once
+per item in the array. To help understand the nuances of array pages, consider
+the following example:
+
+```jsx
+<Page
+  path="/marriages/:marriageIndex/foo"
+  pathParams={{
+    marriageIndex: 'marriages'
+  }}>
+  {({ formData, pathParams }) => (
+    <Field name={`marriages.${pathParams.marriageIndex}.name.first`} />
+  )}
+</Page>
+```
+
+For array pages to work we need a couple core things:
+- The path for the page
+- The data for the current item in the array
+
+To get these pieces, `Page` will accept the `path` and `pathParams` props.
+
+`path` will be the URL within the application and must (for array pages) contain
+path parameters. A path parameter must be the only token in the URL segment
+(between `/`s or at the end of the URL). For example, `/:index/foo` and
+`/:index` are valid path parameters. `/my-:index-item` is an invalid path
+parameter. The name of the path parameter may be anything, but it must start
+with a `:`.
+
+`pathParams` will be an object whose keys map to the path parameters and whose
+values map to the path to the form data. In the above example, the `pathParams`
+will assign `:marriageIndex` in the URL a number based in the current item in
+the `marriages` array.
+
+##### Review page
+To keep Formulate simple, it will not build the review page automatically. It
```

Comment:
I think this is going to be hard to swallow if it's not accompanied by a recommendation from the design system group of a simpler pattern for form review pages. Even with the components listed below, building a review page like the current one is going to be difficult and time consuming, particularly for large forms.

Reply:
Agreed. The review page is the part I'm the least comfortable about for this whole design document. I just don't know how to automate the creation of it without some pretty nasty blows to the developer experience when using the rest of the library. :disappointed: 

The root cause of many of the problems with the current forms library—as I see it—is the centralized configuration for building form elements. That's the main problem I'm trying to solve. But without that centralized configuration, it's really tough to build all the pages _and_ the review mode for each page all at once automatically. So barring that, my next best option I could come up with was to simplify the creation—and customization—of the review page.

If you or anybody else has ideas for how to make that happen, I'd love to hear them. :sweat_smile: